haystack/components/rankers/metadata_grouper.py

```diff
@@ -0,0 +1,190 @@
+    ):  # pylint: disable=too-many-positional-arguments
+        """
+        Sorts the documents within the groups or subgroups based on the 'sort_docs_by' value.
+
+        If 'sort_docs_by' is not provided, the documents are kept in the same order as they were inserted in the groups
+        and subgroups.
+
+        :param document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+        :param document_subgroups: A dictionary with the 'subgroup_by' values as keys and the documents as values.
+        :param no_group: A list of documents without a 'group_by' value.
+        :param ordered_keys: A list of 'group_by' values in the order they were encountered.
+        :param subgroup_ordered_keys: A list of 'subgroup_by' values in the order they were encountered.
+        :returns:
+            A list of documents ordered by the 'sort_docs_by' metadata values.
+        """
+        result_docs = []
+        if self.sort_docs_by and not self.subgroup_by:
+            for key in document_groups:
+                result_docs += sorted(document_groups[key], key=lambda d: d.meta.get(self.sort_docs_by, float("inf")))
+
+        elif self.sort_docs_by and self.subgroup_by:
+            for group in ordered_keys:
+                for subgroup in subgroup_ordered_keys:
+                    result_docs += sorted(
+                        document_subgroups[group][subgroup], key=lambda d: d.meta.get(self.sort_docs_by, float("inf"))
+                    )
+        else:
+            for key in document_groups:
+                result_docs += document_groups[key]
+        for doc in no_group:
+            if doc not in result_docs:
+                result_docs += [doc]
+
+        return result_docs
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document]) -> Dict[str, Any]:
+        """
+        Groups the provided list of documents based on the `group_by` parameter and optionally the `subgroup_by`.
+
+        The output is a list of documents re-ordered based on how they were grouped.
+
+        :param documents: The list of documents to group.
+        :returns:
+            A dictionary with the following keys:
+            - documents: The list of documents ordered by the `group_by` and `subgroup_by` metadata values.
+        """
+        if len(documents) == 0:
+            return {"documents": []}
+
+        # docs based on the 'group_by' value.
+        document_groups, no_group, ordered_keys = self._buck_into_groups(documents)
+
+        # further bucking of the document groups based on the 'subgroup_by' value
+        document_subgroups, subgroup_ordered_keys = self._buck_into_subgroups(document_groups)
+
+        # sort the docs within the groups or subgroups if necessary
+        result_docs = self._sort_docs(
+            document_groups, document_subgroups, no_group, ordered_keys, subgroup_ordered_keys
+        )
+
+        return {"documents": result_docs}
```

Comment:
this could become:

```python
        if not documents:
            return {"documents": []}

        groups = self._group_documents(documents)
        subgroups = self._create_subgroups(groups)
        ordered_docs = self._merge_and_sort(groups, subgroups)

        return {"documents": ordered_docs}
```

Reply:
this misses, as I mentioned above, documents without the specified metadata keys and also collecting the values of those metadata keys which are later used for the final sorting